File: algorithms-python/trees/traverse/preorder_traversal.py
from typing import Deque

from trees.node import Node
from collections import deque

'''
                        1
                     /      \
                    2        3
                  /   \        \
                4        5       6
                                /
                               7           
'''


def preorder2(root: Node) -> None:
    if root is None:
        return
    stack: Deque[Node] = deque()
    stack.append(root)
    while stack:
        node: Node = stack.pop()
        print(node.data, end="\t")
        if node.right is not None:
            stack.append(node.right)
        if node.left is not None:
            stack.append(node.left)


def preorder1(root: Node) -> None:
    if root is None:
        return
    print(root.data, end="\t")
    preorder1(root.left)
    preorder1(root.right)


def main():
    node1: Node = Node(1)
    node2: Node = Node(2)
    node3: Node = Node(3)
    node4: Node = Node(4)
    node5: Node = Node(5)
    node6: Node = Node(6)
    node7: Node = Node(7)
    node1.left = node2
    node1.right = node3
    node2.left = node4
    node2.right = node5
    node3.right = node6
    node6.left = node7
    preorder1(node1)
    print()
    preorder2(node1)


if __name__ == '__main__':
    main()

File: algortihms-java/src/com/algo/tree/traversal/PreOrderTraversal.java
package com.algo.tree.traversal;

import com.algo.tree.Node;

import java.util.Stack;

public class PreOrderTraversal {
    public static void main(String[] args) {
        Node node1 = new Node(1);
        Node node2 = new Node(2);
        Node node3 = new Node(3);
        Node node4 = new Node(4);
        Node node5 = new Node(5);
        Node node6 = new Node(6);
        node1.setLeft(node2);
        node1.setRight(node3);
        node2.setLeft(node4);
        node2.setRight(node5);
        node3.setLeft(node6);
        preOrder1(node1);
        System.out.println("\n");
        preOrder2(node1);

    }

    public static void preOrder1(Node root) {
        if (root == null) {
            return;
        }
        System.out.print(root.getData() + "\t");
        preOrder1(root.getLeft());
        preOrder1(root.getRight());
    }

    private static void preOrder2(Node root) {
        Stack<Node> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            Node node = stack.pop();
            System.out.print(node.getData() + "\t");
            if (node.getRight() != null) {
                stack.push(node.getRight());
            }
            if (node.getLeft() != null) {
                stack.push(node.getLeft());
            }
        }
    }
}
